### PR TITLE
Updating the chemistry tests and docs

### LINKED ISSUE #258: Update all chemistry tests to use FpcTP state vars
Most all chemistry tests (and sub-flowsheet objects) use the 'FTPx' state variables. We want to move towards using the 'FpcTP' state variables because this scheme is more numerically stable, especially for multi-phase flows. Each set of test files should be gone through and updated to use the newer state variable definition. **This will also require changing how each problem is scaled and initialized!!!** Thus, this is a pretty large task. 

Tests to update:
- [x] test_chlorination
- [x] ~test_enrtl_water_pH~
- [x] ~test_pure_water_pH~
- [x] test_recarbonation_process
- [x] test_remineralization 
- [x] ~test_seawater_alkalinity~
- [x] ~test_water_softening~
- [x] **REMOVE** test_phoshorus_removal (made obsolete by #240)
- [x] **REMOVE** test_water_softening_Ca (redundant) 
- [x] **REMOVE** test_water_softening_Mg (redundant)
- [x] **REMOVE** test_enrtl_water_pH_edb (does nothing)

### LINKED ISSUE #272: Update documentation on chemistry/EDB examples/tests
Some of the documentation is outdated or needs to be cleaned up a bit

- [x] Update the 'How to Scale Chemistry' documentation (outdated information)
- [x] Update the example chemistry flowsheets notes/documents 
- [x] Update EDB 'How to' docs / jupyter notebooks to reflect that fact that MongoDB MUST be installed first
- [x] Remove redundant/unused samples from old flowsheet examples

### PR DESCRIPTION
**Pull request by aladshaw3**
## Fixes/Addresses:

#258 & #272

Several files are touched in this PR. Most of this is changes to documentation or the removal of outdated/redundant examples.

## Summary/Motivation:

PR is to remove outdated tests and update some tests with FpcTP state vars. 

## Changes proposed in this PR:
- Removal of outdated/redundant tests
- Updating methods used in tests
- Updating documentation to reflect new chemistry scaling standards
- Cleaning up docs
- Removing multiple duplicate flowsheet examples under the chemistry folder

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: watertap/examples/chemistry/chem_scaling_utils.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    for phase in unit.control_volume.properties_in[0.0].enth_mol_phase:
+        val = abs(value(unit.control_volume.properties_in[0.0].enth_mol_phase[phase].expr))
+        if val >= max:
+            max = val
+        if val <= min:
+            val = min
+        iscale.set_scaling_factor(unit.control_volume.properties_in[0.0]._enthalpy_flow_term[phase], 10/val)
+        iscale.set_scaling_factor(unit.control_volume.properties_out[0.0]._enthalpy_flow_term[phase], 10/val)
+
+    iscale.constraint_scaling_transform(unit.control_volume.enthalpy_balances[0.0], 10/max)
```
Comment (bknueven) [inline watertap/examples/chemistry/chem_scaling_utils.py]:
```suggestion
    max_enth_mol_phase = 1
    min_scale = 1
    for phase in unit.control_volume.properties_in[0.0].enth_mol_phase:
        val = max(min_scale, abs(value(unit.control_volume.properties_in[0.0].enth_mol_phase[phase].expr)))
        max_enth_mol_phase = max(val, max_enth_mol_phase)
        iscale.set_scaling_factor(unit.control_volume.properties_in[0.0]._enthalpy_flow_term[phase], 10/val)
        iscale.set_scaling_factor(unit.control_volume.properties_out[0.0]._enthalpy_flow_term[phase], 10/val)

    iscale.constraint_scaling_transform(unit.control_volume.enthalpy_balances[0.0], 10/max_enth_mol_phase)
```

File: watertap/examples/chemistry/chem_scaling_utils.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+## Helper function for setting scaling factors for the mass balance for FTPx state vars
+def _set_mat_bal_scaling_FTPx(unit, min=1e-3):
+    # For species
+    for i in unit.control_volume.properties_out[0.0].mole_frac_phase_comp:
+        # i[0] = phase, i[1] = species
+        if unit.inlet.mole_frac_comp[0, i[1]].value > min:
+            scale = unit.inlet.mole_frac_comp[0, i[1]].value
+        else:
+            scale = min
```
Comment (bknueven) [inline watertap/examples/chemistry/chem_scaling_utils.py]:
```suggestion
def _set_mat_bal_scaling_FTPx(unit, min_mole_frac_comp=1e-3):
    # For species
    for i in unit.control_volume.properties_out[0.0].mole_frac_phase_comp:
        # i[0] = phase, i[1] = species
        scale = max(min_mole_frac_comp, unit.inlet.mole_frac_comp[0, i[1]].value)
```

File: watertap/examples/chemistry/chem_scaling_utils.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+## Helper function for setting scaling factors for the mass balance for FpcTP state vars
+def _set_mat_bal_scaling_FpcTP(unit, min=1e-3):
+    # For species
+    for i in unit.control_volume.properties_out[0.0].mole_frac_phase_comp:
+        # i[0] = phase, i[1] = species
+        if unit.inlet.flow_mol_phase_comp[0, i[0], i[1]].value > min:
+            scale = unit.inlet.flow_mol_phase_comp[0, i[0], i[1]].value
+        else:
+            scale = min
```
Comment (bknueven) [inline watertap/examples/chemistry/chem_scaling_utils.py]:
```suggestion
def _set_mat_bal_scaling_FpcTP(unit, min_flow_mol_phase_comp=1e-3):
    # For species
    for i in unit.control_volume.properties_out[0.0].mole_frac_phase_comp:
        # i[0] = phase, i[1] = species
        scale = max(min_flow_mol_phase_comp, unit.inlet.flow_mol_phase_comp[0, i[0], i[1]].value)
```

File: watertap/examples/chemistry/chem_scaling_utils.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        # i[0] = time, i[1] = reaction
+
+        # Grab the 'k_eq_ref' value from the thermo config
+        scale = max(min_scale, thermo_config["inherent_reactions"][i[1]]["parameter_data"]["k_eq_ref"][0])
+        iscale.set_scaling_factor(unit.control_volume.inherent_reaction_extent[0.0,i[1]], 10/scale)
+        iscale.constraint_scaling_transform(unit.control_volume.properties_out[0.0].
+                inherent_equilibrium_constraint[i[1]], 0.1)
+
+## Helper function for setting scaling factors for rate reactions
+def _set_rate_rxn_scaling(rxn_params, unit, min_scale=1e-3):
```
Comment (bknueven) [inline watertap/examples/chemistry/chem_scaling_utils.py:60]: `min_scale` isn't used. Should it be?
```suggestion
def _set_rate_rxn_scaling(rxn_params, unit):
```

File: watertap/examples/chemistry/chem_scaling_utils.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+            unit.control_volume.reactions[0.0].equilibrium_constraint[i[1]], 0.1)
+
+## Helper function for setting scaling factors for inherent reactions
+def _set_inherent_rxn_scaling(unit, thermo_config, min_scale=1e-3):
+    #Add scaling factors for reactions (changes depending on if it is a log form or not)
+    for i in unit.control_volume.inherent_reaction_extent_index:
+        # i[0] = time, i[1] = reaction
+
+        # Grab the 'k_eq_ref' value from the thermo config
+        scale = max(min_scale, thermo_config["inherent_reactions"][i[1]]["parameter_data"]["k_eq_ref"][0])
```
Comment (bknueven) [inline watertap/examples/chemistry/chem_scaling_utils.py]:
```suggestion
def _set_inherent_rxn_scaling(unit, thermo_config, min_k_eq_ref=1e-3):
    #Add scaling factors for reactions (changes depending on if it is a log form or not)
    for i in unit.control_volume.inherent_reaction_extent_index:
        # i[0] = time, i[1] = reaction
        # Grab the 'k_eq_ref' value from the thermo config
        scale = max(min_k_eq_ref, thermo_config["inherent_reactions"][i[1]]["parameter_data"]["k_eq_ref"][0])
```

File: watertap/examples/chemistry/chem_scaling_utils.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+                    rxn_params.component("reaction_"+rid).eps.value = 1e-16*factor
+
+## Helper function for setting scaling factors for equilibrium reactions
+def _set_equ_rxn_scaling(unit, rxn_config, min_scale=1e-3):
+    #Add scaling factors for reactions (changes depending on if it is a log form or not)
+    for i in unit.control_volume.equilibrium_reaction_extent_index:
+        # i[0] = time, i[1] = reaction
+
+        # Grab the 'k_eq_ref' value from the reaction config
+        scale = max(min_scale, rxn_config["equilibrium_reactions"][i[1]]["parameter_data"]["k_eq_ref"][0])
```
Comment (bknueven) [inline watertap/examples/chemistry/chem_scaling_utils.py]:
```suggestion
def _set_equ_rxn_scaling(unit, rxn_config, min_k_eq_ref=1e-3):
    #Add scaling factors for reactions (changes depending on if it is a log form or not)
    for i in unit.control_volume.equilibrium_reaction_extent_index:
        # i[0] = time, i[1] = reaction
        # Grab the 'k_eq_ref' value from the reaction config
        scale = max(min_k_eq_ref, rxn_config["equilibrium_reactions"][i[1]]["parameter_data"]["k_eq_ref"][0])
```

File: watertap/examples/chemistry/chem_scaling_utils.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+            # Grab the 'k_eq_ref' value from the reaction config
+            scale = rxn_config["equilibrium_reactions"][rid]["parameter_data"]["k_eq_ref"][0]
+
+            # NOTE: ONLY certain functions have an eps value that we need to set
+            if (hasattr(rxn_params.component("reaction_"+rid),"eps")):
+                # highest allowable value for setting eps based on k_eq_ref
+                if scale < 1e-16:
+                    rxn_params.component("reaction_"+rid).eps.value = scale*factor
+                else:
+                    rxn_params.component("reaction_"+rid).eps.value = 1e-16*factor
```
Comment (bknueven) [inline watertap/examples/chemistry/chem_scaling_utils.py]: Should `1e-16` also be settable from the function?
```suggestion
                # highest allowable value for setting eps based on k_eq_ref
                rxn_params.component("reaction_"+rid).eps.value = min(1e-16, scale)*factor
```

**Review by bknueven — COMMENTED** [review]
Since we're consolidating the chemical scaling utility functions, and having worked with them before, I have several suggestions to improve their readability. Mainly to be more explicit in names and use the built-in `min`/`max` functions to communicate intent.

File: watertap/examples/chemistry/chem_scaling_utils.py (hunk 7 above)
Comment (aladshaw3) [reply watertap/examples/chemistry/chem_scaling_utils.py]: I guess we 'could' let the 1e-16 be settable. I will consider changing that.

File: watertap/examples/chemistry/chem_scaling_utils.py (hunk 4 above)
Comment (aladshaw3) [reply watertap/examples/chemistry/chem_scaling_utils.py:60]: I haven't yet fully worked out how I actually want to apply scaling for this type of reaction. What I have right now works, but I have done only very limited testing with it. I anticipate it needing some 'min_scale' arg later, so I put one in the function call. I may add a TODO statement here as a reminder.

**Review by bknueven — APPROVED** [review]
I like the cleanup of the existing code.

Question: Is there any reason not to make the functions in `watertap.examples.chemistry.chem_scaling_utils` public, and just document the functions? Presently we include their implementation in the documentation, which seems a bit unusual to me, and has the potential for the `chem_scaling_utils` and recommended scaling implementation to diverge at some point (as is already happening within this PR).

**Comment by aladshaw3** [comment]
@bknueven  Thanks for reviewing the code changes. I had thought about moving the scaling utils for the chemistry stuff to a more permanent location, but given that our state of scaling is still a little in flux, I decided not to. I have a feeling more things will change as more testing is done, especially once we start encapsulating some of the chemistry into there own unit models and/or sub-flowsheets. 

The documentation on scaling is meant primarily to be a guide for users/developers because of how difficult scaling can be to apply for specifically the chemistry. Thus, the documentation is mostly meant to provide a starting point for people getting started (and for me to remember what I did and why). 

I think it is a good idea to eventually move scaling utils to a more permanent location, but think we should finalize all the changes we make to how we scale and solve WaterTAP models first.

**Review by srikanthallu — APPROVED** [review]

**Pull request merged**